src/plans/count.jl

```diff
@@ -0,0 +1,200 @@
+"""
+     CountObjective{E,O<:AbstractManifoldObjective,I<:Integer} <: AbstractManifoldObjective{E}
+
+A wrapper for any [`AbstractManifoldObjective`](@ref) to count different calls to parts of
+the objective.
+
+# Fields
+
+* `counts` a dictionary of symbols mapping to integers keeping the counted values
+* `objective` the wrapped objective
+
+# Supported Symbols
+
+| Symbol                    | Counts calls to                      | Comment                      |
+| ------------------------- | -----------------------------        | ------------------           |
+| `:Cost`                   | [`get_cost`](@ref)                   |                              |
+| `:Gradient`               | [`get_gradient`](@ref)               |                              |
+| `:Hessian`                | [`get_hessian`](@ref)                |                              |
+| `:Preconditioner`         | [`get_preconditioner`](@ref)         |                              |
+| `:Constraints`            | [`get_constraints`](@ref)            |                              |
+| `:EqualityConstraint`     | [`get_equality_constraint`](@ref)    | requires vector of counters  |
+| `:EqualityConstraints`    | [`get_equality_constraints`](@ref)   | does not count single access |
+| `:InequalityConstraint`   | [`get_inequality_constraint`](@ref)  | requires vector of counters  |
+| `:InequalityConstraints`  | [`get_inequality_constraints`](@ref) | does not count single access |
+
+# Constructors
+
+    CountObjective(objective::AbstractManifoldObjective, counts::Dict{Symbol, <:Integer})
+
+Initialise the `CountObjective` to wrap `objective` initializing the set of counts
+
+    CountObjective(M::AbtractManifold, objective::AbstractManifoldObjective, count::AbstractVecor{Symbol}, init=0)
+
+Count function calls on `objective` using the symbols in `count` initialising all entries to `init`.
+"""
+struct CountObjective{E,O<:AbstractManifoldObjective,I<:Integer} <:
+       AbstractManifoldObjective{E}
+    counts::Dict{Symbol,I}
+    objective::O
+end
+function CountObjective(
+    ::AbstractManifold, objective::O, counts::Dict{Symbol,I}
+) where {E<:AbstractEvaluationType,I<:Integer,O<:AbstractManifoldCostObjective{E}}
+    return CountObjective{E,O,I}(counts, objective)
+end
+function CountObjective(
+    ::AbstractManifold, objective::O, count::AbstractVector{Symbol}, init::I=0
+) where {E<:AbstractEvaluationType,I<:Integer,O<:AbstractManifoldCostObjective{E}}
+    return CountObjective(objective, Dict([symbol => init for symbol in count]))
+end
+dispatch_objective_decorator(::CountObjective) = Val(true)
+
+function _count_if_exists(co::CountObjective, s::Symbol)
+    return haskey(co.counts, s) && (co.counts[s] += 1)
+end
+function _count_if_exists(co::CountObjective, s::Symbol, i)
+    return haskey(co.counts, s) &&
+           length(i) == ndims(co.counts[s]) &&
+           all(i .<= size(co.counts[s])) &&
+           (co.counts[s][i] += 1)
```

Comment:
`co.counts[s]` is an integer, how is it supposed to work?

Reply:
This is still a bit work in progress, in this case it is not just an integer, that is the case above.

The application in mind is the following: CPPA has a whole vector of proxies – the same for the inequality functions, equality functions or elements for stochastic gradient.
In these cases, we want to (hopefully!) count the individual calls to the proxes. That is what this aims for. But I am not yet done sketching this and have written the tests – initialising these “group counts” might be tricky.

Reply:
I see, then it could be a bit tricky to implement.